fix loading 100 rows in db import preview

smierz opened the pull request:
fixes #7499 

**Changes proposed in this pull request**
- Currently the `end` flag that should signal the retrieval of the last batch from the db is set to `true` after loading first batch (because `dbRows.size() < batchSize + 1` is always true). Fixed by setting it to `dbRows.size() < batchSize` so the `end` flag is only set for the last batch.

- Additionally `dbRows` is checked for null in the previous if-statement, but not before setting the `end` flag this way. Fixed by putting a guard statement in directly after loading `dbRows`.

- The `end` parameter is set within `getRowsOfCells` and every call to this method after setting the flag to true should not result in additional db queries. That's why it should be checked first thing within the method instead of outside of it. Moved the flag check.

tfmorris commented on extensions/database/src/com/google/refine/extension/database/DBQueryResultImportReader.java:
```diff
@@ -208,14 +208,12 @@ private List<List<Object>> getRowsOfCells(int startRow) throws IOException, Data
 
         }
+
         end = dbRows.size() < batchSize + 1;
```
This looks suspicious. Should the `+ 1` be removed to align with the other change you made?

tfmorris reviewed (commented):
I should have looked at the code before offering an opinion on refactoring. Now I'm torn. One the one hand, it's basically a one line fix, which I'm reluctant to delay, but on the other hand, you're having to generate additional test code due to the duplication and the duplication means that there's a risk additional bugs in the parallel code. Having said that, I suspect that the refactoring will be non-trivial because the current control flow is a little messy. I don't understand why it's pre-fetching rows when there doesn't seem to be an obvious performance advantage to the extra complexity.

If you decide to stick with the current approach, the most important comments concern the commented out `assert` and the potential off-by-one problem.

smierz commented on extensions/database/tests/src/com/google/refine/extension/database/DBQueryResultPreviewReaderTest.java:
```diff
@@ -0,0 +1,158 @@
+import com.google.refine.extension.database.model.DatabaseRow;
+import com.google.refine.importing.ImportingJob;
+
+public class DBQueryResultPreviewReaderTest {
```
the duplicate code is only temporary (preview reader and its tests will be removed), that's why I did not invest time into refactoring

smierz commented:
There were some minor things I noticed and refactored:
* only changed one occurrence of returning an empty list, but there was one more 
-> [replace with ref to empty list](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/d6ac840a236fe2a888a14ea8ef1a5d61a4799914)
* we can construct the header in a one-liner
-> [refactor construction of header into one-liner](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/4798e6d3486d9a54673810488bdb03a4a74c1fc7)
* (review comment) switched to lazy eval of debugging parameter using supplier  
-> [switch to lazy eval of logging parameter](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/6389600ebc176555218eeaf9edfaea5eb00cd89b)
* renamed the `end` parameter to adhere to Java convention and also to make it clearer what it flags 
-> [rename boolean end for clarity and adherence to convention](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/7463b251ffd0bf9eae8a83cb3ca623a52cb387d4)
* switched to 0-based log statements for the batch intervals 
-> [switch to 0-based logging for batches](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/bb959f5748f1e2182c7f60066ec3a8d1542e3b68)

Additionally I fixed a bunch of things IntelliJ flagged:
* [make parameters final](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/5add945a8b68635dc9005ac7bb5dc90524cb02ec)
* [fix logging parameter](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/d43c687a38aecc1ce39c32dcd9469f1ce6e91ca8)
* [rm IOException from method signature as it is never thrown](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/c1cc3ad0ef646e93c928cfb7c9b54f6cefff48b6)

smierz commented on extensions/database/src/com/google/refine/extension/database/DBQueryResultImportReader.java:
```diff
@@ -147,11 +150,10 @@ public List<Object> getNextRowOfCells() throws IOException {
 
         } catch (DatabaseServiceException e) {
+            // rethrow exception as IOException
             logger.error("DatabaseServiceException::{}", e);
```
changed it in [fix parameter in error log](https://github.com/OpenRefine/OpenRefine/pull/7507/commits/2db84a5a6e296a8221087d2d0982b427eb3796f8)

tfmorris reviewed (approved):
Looks great. Thanks for all the work!
